sequencer/src/state.rs

```diff
@@ -308,6 +321,228 @@ fn validate_and_apply_header(
+    block_height: u64,
+}
+
+#[async_trait]
+impl<'a> StateCatchup for SqlStateCatchup<'a> {
+    async fn fetch_accounts(
+        &self,
+        _view: ViewNumber,
+        _fee_merkle_tree_root: FeeMerkleCommitment,
+        accounts: Vec<FeeAccount>,
+    ) -> anyhow::Result<Vec<AccountQueryData>> {
+        let mut ret = vec![];
+        for account in accounts {
+            let block_height = self.block_height;
+
+            let proof = self
+                .db
+                .get_path(
+                    Snapshot::<SeqTypes, FeeMerkleTree, 256>::Index(block_height),
+                    account,
+                )
+                .await?;
+
+            match proof.proof.first().context("empty proof")? {
+                MerkleNode::Leaf { pos, elem, .. } => {
+                    ret.push(AccountQueryData {
+                        balance: elem.0,
+                        proof: FeeAccountProof {
+                            account: (*pos).into(),
+                            proof: FeeMerkleProof::Presence(proof),
+                        },
+                    });
+                }
+
+                MerkleNode::Empty => {
+                    ret.push(AccountQueryData {
+                        balance: 0_u64.into(),
+                        proof: FeeAccountProof {
+                            account: account.into(),
+                            proof: FeeMerkleProof::Absence(proof),
+                        },
+                    });
+                }
+                _ => {
+                    bail!("Invalid proof");
+                }
+            }
+        }
+        Ok(ret)
+    }
+
+    async fn remember_blocks_merkle_tree(
+        &self,
+        _view: ViewNumber,
+        mt: &mut BlockMerkleTree,
+    ) -> anyhow::Result<()> {
+        let bh = self.block_height;
+
+        if bh == 0 {
+            return Ok(());
+        }
+
+        let proof = self
+            .db
+            .get_path(Snapshot::<SeqTypes, BlockMerkleTree, 3>::Index(bh), bh - 1)
+            .await
+            .unwrap();
+
+        match proof.proof.first().context("proof is empty")? {
+            MerkleNode::Leaf { pos, elem, .. } => mt
+                .remember(pos, elem, proof.clone())
+                .context("failed to remember proof"),
+            _ => bail!("invalid proof"),
+        }
+    }
+}
+
+pub async fn update_state_storage(
+    storage: Arc<RwLock<SqlDataSource<SeqTypes, Provider>>>,
+    instance: NodeState,
+) -> anyhow::Result<()> {
+    // get last saved merklized state
+
+    let storage_read = storage.read().await;
+    let last_height = storage_read.get_last_state_height().await?;
+
+    let mut leaves = storage_read.subscribe_leaves(last_height).await;
+
+    while let Some(leaf) = leaves.next().await {
+        let leaf = leaf.leaf();
+        let header = leaf.get_block_header();
+        let validated_state = ValidatedState::from_header(header);
```

Comment:
This is the parent state we're applying onto so we need to create it from the parent header